src/D3MHub.sol

```diff
@@ -221,121 +219,123 @@ contract D3MHub {
+            vat.frob(ilk, address(_pool), address(_pool), address(this), int256(fixInk), 0); // Lock it
+            ink += fixInk;
+            emit Fees(ilk, fixInk);
+        }
+        if (art < ink) { // If there was permissionless DAI paid or fees added as collateral
+            address _vow = vow;
+            uint256 fixArt = ink - art; // Amount of fees + permissionless DAI paid we will now transform to debt
+            delete art;
+            Art += fixArt;
+            require(Art <= MAXINT256, "D3MHub/overflow");
+            vat.suck(_vow, _vow, fixArt * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(_pool), address(_pool), _vow, 0, int256(fixArt)); // Generating the debt
         }
 
-        // To save gas you can bring the fees back with the unwind
-        uint256 total = amount + fees;
-        _pool.withdraw(total);
-        daiJoin.join(address(this), total);
+        // Determine if it needs to unwind or wind
 
-        // normalized debt == erc20 DAI to pool (Vat rate for D3M ilks fixed to 1 RAY)
+        uint256 Line = vat.Line();
+        uint256 debt = vat.debt();
 
-        if (mode == Mode.NORMAL) {
-            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, fees * RAY);
-        } else if (mode == Mode.D3M_CULLED) {
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
+        uint256 toUnwind;
+        if (ilks[ilk].tic != 0 || !ilks[ilk].plan.active()) { // If D3M is caged (but not culled) or plan is not active
+            toUnwind = maxWithdraw; // we want to unwind the most we can
         } else {
-            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
-            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
-            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
-            vat.slip(ilk, address(_end), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
-        }
+            if (Art > lineWad) {
+                unchecked {
+                    toUnwind = Art - lineWad; // checks if we need to unwind due ilk debt ceiling
+                }
+            }
+            if (debt > Line) {
+                unchecked {
+                    toUnwind = _max(toUnwind, _divup(debt - Line, RAY)); // checks if we need to unwind due global debt ceiling
+                }
+            }
 
-        emit Unwind(ilk, amount, fees);
-    }
+            // Determine if it needs to unwind due plan
+            uint256 targetAssets = ilks[ilk].plan.getTargetAssets(currentAssets);
```

Comment:
`getTargetAssets` appears twice in the code (and is called twice if we end up in the `wind` path).
Can we store it in a local var that's declared near `toUnwind`? (tried it locally and did not encounter stack too deep problems).

Reply:
Fixed!